compiler/noirc_frontend/src/monomorphization/mod.rs

```diff
@@ -224,30 +255,43 @@ impl<'interner> Monomorphizer<'interner> {
         main_meta.function_signature()
     }
 
+    fn compile_main_debug(
+        &mut self,
+        main_id: node_interner::FuncId,
+        field_names: &HashMap<u32, String>,
+    ) -> FunctionSignature {
+        self.debug_field_names = field_names.clone();
+        self.compile_main(main_id)
+    }
+
     fn function(&mut self, f: node_interner::FuncId, id: FuncId) {
         if let Some((self_type, trait_id)) = self.interner.get_function_trait(&f) {
             let the_trait = self.interner.get_trait(trait_id);
             the_trait.self_type_typevar.force_bind(self_type);
         }
 
-        let meta = self.interner.function_meta(&f);
-        let modifiers = self.interner.function_modifiers(&f);
-        let name = self.interner.function_name(&f).to_owned();
+        let (meta_parameters, unconstrained, body_expr_id, name, return_type) = {
+            let meta = self.interner.function_meta(&f).clone();
+            let modifiers = self.interner.function_modifiers(&f).clone();
+            let name = self.interner.function_name(&f).to_string();
 
-        let body_expr_id = *self.interner.function(&f).as_expr();
-        let body_return_type = self.interner.id_type(body_expr_id);
-        let return_type = self.convert_type(match meta.return_type() {
-            Type::TraitAsType(..) => &body_return_type,
-            _ => meta.return_type(),
-        });
-
-        let parameters = self.parameters(&meta.parameters);
+            let body_expr_id = *self.interner.function(&f).as_expr();
+            let body_return_type = self.interner.id_type(body_expr_id);
+            let return_type = self.convert_type(match meta.return_type() {
+                Type::TraitAsType(_, _, _) => &body_return_type,
+                _ => meta.return_type(),
+            });
+            let unconstrained = modifiers.is_unconstrained
+                || matches!(modifiers.contract_function_type, Some(ContractFunctionType::Open));
+            (meta.parameters, unconstrained, body_expr_id, name, return_type)
+        };
 
-        let body = self.expr(body_expr_id);
-        let unconstrained = modifiers.is_unconstrained
-            || matches!(modifiers.contract_function_type, Some(ContractFunctionType::Open));
+        let function = {
+            let parameters = self.parameters(&meta_parameters).clone();
+            let body = self.expr(body_expr_id).clone();
+            ast::Function { id, name, parameters, body, return_type, unconstrained }
+        };
```

Comment:
What changed about this section? It looks like we're introducing a lot more `clone`s but I don't see changes beyond that. I know the interner is mutable now but I think we should be able to avoid most of these clones still.

Reply:
You're right. We only need to `clone()` the `meta` variable. The rest can remain the same. Maybe something changed, but I distinctly remember fighting the borrow checker on this one. But maybe it was just my inexperience. Thanks!

Reply:
The other required change was in the order of some of the statements, due to borrow checker constraints.